cdap-master/src/main/java/co/cask/cdap/data/tools/flow/FlowQueuePendingCorrector.java

```diff
@@ -0,0 +1,229 @@
+import co.cask.cdap.data2.transaction.queue.QueueAdmin;
+import co.cask.cdap.data2.transaction.queue.hbase.HBaseQueueAdmin;
+import co.cask.cdap.data2.transaction.queue.hbase.HBaseQueueClientFactory;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtil;
+import co.cask.cdap.data2.util.hbase.HBaseTableUtilFactory;
+import co.cask.cdap.internal.app.runtime.flow.FlowUtils;
+import co.cask.cdap.metrics.guice.MetricsClientRuntimeModule;
+import co.cask.cdap.proto.Id;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.util.concurrent.AbstractIdleService;
+import com.google.gson.Gson;
+import com.google.inject.AbstractModule;
+import com.google.inject.Guice;
+import com.google.inject.Inject;
+import com.google.inject.Injector;
+import com.google.inject.Singleton;
+import org.apache.commons.cli.BasicParser;
+import org.apache.commons.cli.CommandLine;
+import org.apache.commons.cli.CommandLineParser;
+import org.apache.commons.cli.HelpFormatter;
+import org.apache.commons.cli.Option;
+import org.apache.commons.cli.Options;
+import org.apache.commons.cli.ParseException;
+import org.apache.twill.kafka.client.KafkaClientService;
+import org.apache.twill.zookeeper.ZKClientService;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Corrects the "queue.pending" metric emitted in {@link co.cask.cdap.internal.app.runtime.flow.FlowletProgramRunner}.
+ */
+public class FlowQueuePendingCorrector extends AbstractIdleService {
+
+  private static final Gson GSON = new Gson();
+
+  private final MetricsCollectionService metricsCollectionService;
+  private final MetricStore metricStore;
+  private final KafkaClientService kafkaClientService;
+  private final HBaseQueueDebugger queueDebugger;
+  private final ZKClientService zkClientService;
+
+  @Inject
+  public FlowQueuePendingCorrector(HBaseQueueDebugger queueDebugger, ZKClientService zkClientService,
+                                   MetricsCollectionService metricsCollectionService, MetricStore metricStore,
+                                   KafkaClientService kafkaClientService) {
+    this.queueDebugger = queueDebugger;
+    this.zkClientService = zkClientService;
+    this.metricsCollectionService = metricsCollectionService;
+    this.metricStore = metricStore;
+    this.kafkaClientService = kafkaClientService;
+  }
+
+  public void run(Id.Flow flow, String producerFlowlet, String consumerFlowlet,
+                  String flowletQueue) throws Exception {
+
+    QueueName queueName = QueueName.fromFlowlet(flow, producerFlowlet, flowletQueue);
+    long consumerGroupId = FlowUtils.generateConsumerGroupId(flow, consumerFlowlet);
+    HBaseQueueDebugger.QueueStatistics stats = queueDebugger.scanQueue(queueName, consumerGroupId);
+    long correctQueuePendingValue = stats.getUnprocessed() + stats.getProcessedAndNotVisible();
```

Comment:
Because this should only run when the flow is stopped, there should not be any invisible processed entries. Can you add a check that it is zero and print a warning if it is not?